### PR TITLE
Organization CRUD in GraphQL

### PR DESCRIPTION
**Pull request by moncapitaine**
#909

### CODE REVIEW

File: packages/backend/src/graphql/organization/mutations/organizationCreateMutationField.test.ts [hunk 1]
```diff
@@ -0,0 +1,65 @@
+import { gql } from 'apollo-server-core'
+import { GraphQLError } from 'graphql'
+
+import { PrismaClient } from '@progwise/timebook-prisma'
+
+import { getTestServer } from '../../../getTestServer'
+
+const orgCreateMutation = gql`
```
Comment (inpaceee) [inline packages/backend/src/graphql/organization/mutations/organizationCreateMutationField.test.ts:8]: would more descriptive name be too large?
```organizationCreateMutation```

File: packages/backend/src/graphql/organization/mutations/organizationCreateMutationField.test.ts [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+    variables: { data: { title: null, address: 'Teststr' } },
+  })
+
+  expect(response.data).toBeUndefined()
+  expect(response.errors?.at(0)?.message).toEqual(
+    'Variable "$data" got invalid value null at "data.title"; Expected non-nullable type "String!" not to be null.',
+  )
+})
+
+it('should throw an error without title', async () => {
```
Comment (inpaceee) [inline packages/backend/src/graphql/organization/mutations/organizationCreateMutationField.test.ts:55]:
```suggestion
it('should create a new ogranization', async () => {
```

File: packages/backend/src/graphql/organization/mutations/organizationDeleteMutationField.ts
Comment (inpaceee) [inline packages/backend/src/graphql/organization/mutations/organizationDeleteMutationField.ts:1]: Is this mutation archiving an organization, not deleting?

File: packages/backend/src/graphql/organization/mutations/organizationUpdateMutationField.ts [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+  t.prismaField({
+    type: 'Organization',
+    description: 'Update an organization',
+    authScopes: async (_source, { id }) => ({ isAdminByOrganization: id.toString() }),
+    args: {
+      id: t.arg.id(),
+      data: t.arg({ type: OrganizationInput }),
+    },
+    resolve: async (query, _source, { id, data: { title, address } }, context) => {
+      const now = new Date()
```
Comment (inpaceee) [inline packages/backend/src/graphql/organization/mutations/organizationUpdateMutationField.ts:15]: Is this needed?

File: packages/backend/src/graphql/organization/mutations/organizationUpdateMutationField.ts [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+builder.mutationField('organizationUpdate', (t) =>
+  t.prismaField({
+    type: 'Organization',
+    description: 'Update an organization',
+    authScopes: async (_source, { id }) => ({ isAdminByOrganization: id.toString() }),
+    args: {
+      id: t.arg.id(),
+      data: t.arg({ type: OrganizationInput }),
+    },
+    resolve: async (query, _source, { id, data: { title, address } }, context) => {
```
Comment (inpaceee) [inline packages/backend/src/graphql/organization/mutations/organizationUpdateMutationField.ts:14]:
```suggestion
    resolve: async (query, _source, { id, data: { title, address } }, _context) => {
```

File: packages/backend/src/graphql/organization/queries/organizationQueryFields.test.ts [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+import { GraphQLError } from 'graphql'
+
+import { PrismaClient } from '@progwise/timebook-prisma'
+
+import { getTestServer } from '../../../getTestServer'
+
+const prisma = new PrismaClient()
+
+const organizationsQuery = gql`
+  query orgs {
```
Comment (inpaceee) [inline packages/backend/src/graphql/organization/queries/organizationQueryFields.test.ts:13]:
```suggestion
  query organizations {
```

File: packages/backend/src/graphql/organization/queries/organizationQueryFields.test.ts [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+
+  await prisma.organizationMembership.create({
+    data: {
+      organizationId: '100',
+      userId: '1',
+    },
+  })
+})
+
+it('user is not signed in', async () => {
```
Comment (inpaceee) [inline packages/backend/src/graphql/organization/queries/organizationQueryFields.test.ts:48]:
```suggestion
it('should throw an error if user is not signed in', async () => {
```

File: packages/backend/src/graphql/organization/queries/organizationQueryFields.test.ts [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+  })
+})
+
+describe('users without orgs', () => {
+  it('should receive no orgs', async () => {
+    const testServer = getTestServer({ userId: '2' })
+    const response = await testServer.executeOperation({ query: organizationsQuery })
+    expect(response.data?.organizations).toEqual([])
+  })
+})
```
Comment (inpaceee) [inline packages/backend/src/graphql/organization/queries/organizationQueryFields.test.ts:70]: is ```describe``` block here for future tests?

File: packages/backend/src/prisma/migrations/20240503135711_organization_metadata/migration.sql [hunk 8]
```diff
@@ -0,0 +1,10 @@
+/*
+  Warnings:
+
+  - Added the required column `updatedAt` to the `Organization` table without a default value. This is not possible if the table is not empty.
+
+*/
+-- AlterTable
+ALTER TABLE "Organization" ADD COLUMN     "archivedAt" TIMESTAMP(3),
+ADD COLUMN     "createdAt" TIMESTAMP(3) NOT NULL DEFAULT CURRENT_TIMESTAMP,
+ADD COLUMN     "updatedAt" TIMESTAMP(3) NOT NULL;
```
Comment (inpaceee) [inline packages/backend/src/prisma/migrations/20240503135711_organization_metadata/migration.sql:10]: Shouldn't this have a default value, as warning suggests?

File: packages/backend/src/prisma/schema.prisma [hunk 9]
```diff
@@ -18,6 +18,26 @@ generator pothos {
...
+
+  organizationMemberships OrganizationMembership[]
+}
+
+model OrganizationMembership {
+  userId         String
+  organizationId String
+
+  @@id([userId, organizationId])
+  organization Organization @relation(fields: [organizationId], references: [id], onDelete: Cascade)
```
Comment (inpaceee) [inline packages/backend/src/prisma/schema.prisma:38]: Shouldn't this have user model relation?
```user User @relation(fields: [userId], references: [id], onDelete: Cascade)```

**Review by inpaceee — APPROVED** [review]

File: packages/backend/src/prisma/schema.prisma [hunk 10]
```diff
@@ -18,6 +18,26 @@ generator pothos {
...
+  createdAt  DateTime  @default(now())
+  updatedAt  DateTime  @updatedAt
+  archivedAt DateTime?
+
+  organizationMemberships OrganizationMembership[]
+}
+
+model OrganizationMembership {
+  userId         String
+  organizationId String
```
Comment (PascalHelbig) [inline packages/backend/src/prisma/schema.prisma:35]: Add `createdAt` and `updatedAt`

File: apps/web/frontend/components/organizationForm/archiveOrUnarchiveOrganizationButton/archiveOrUnarchiveOrganizationButton.tsx [hunk 11]
```diff
@@ -0,0 +1,27 @@
...
+    id
+    isArchived
+    ...UnarchiveOrganizationButton
+    ...ArchiveOrganizationButton
+  }
+`)
+
+interface ArchiveOrUnarchiveOrganizationButtonProps {
+  organization: FragmentType<typeof ArchiveOrUnarchiveOrganizationButtonFragment>
+  disabled: boolean
```
Comment (PascalHelbig) [inline apps/web/frontend/components/organizationForm/archiveOrUnarchiveOrganizationButton/archiveOrUnarchiveOrganizationButton.tsx:16]: where is this used?

File: apps/web/frontend/components/organizationForm/organizationForm.tsx [hunk 12]
```diff
@@ -0,0 +1,92 @@
...
+          label="Name"
+          type="text"
+          disabled={isSubmitting}
+          readOnly={isOrganizationFormReadOnly}
+          {...register('title')}
+          placeholder="Enter organization name"
+          size={30}
+          errorMessage={errors.title?.message}
+          isDirty={isDirty}
+        />
```
Comment (PascalHelbig) [inline apps/web/frontend/components/organizationForm/organizationForm.tsx:62]: Address input is missing

File: apps/web/frontend/components/organizationForm/organizationForm.tsx [hunk 13]
```diff
@@ -0,0 +1,92 @@
...
+  const isNewOrganization = !organization
+  const isOrganizationFormReadOnly = !organization?.canModify && !isNewOrganization
+  return (
+    <div className="mt-4 flex flex-wrap items-start gap-2">
+      <form onSubmit={handleSubmit(handleSubmitHelper)} className="contents" id="organization-form">
+        {isNewOrganization ? (
+          <PageHeading>Create a new organization</PageHeading>
+        ) : (
+          <PageHeading>{isOrganizationFormReadOnly ? 'View' : 'Edit'} organization</PageHeading>
+        )}
```
Comment (PascalHelbig) [inline apps/web/frontend/components/organizationForm/organizationForm.tsx:51]:
```suggestion
  const isOrganizationFormReadOnly = !!organization && !organization.canModify
  return (
    <div className="mt-4 flex flex-wrap items-start gap-2">
      <form onSubmit={handleSubmit(handleSubmitHelper)} className="contents" id="organization-form">
        {organization ? (
          <PageHeading>{isOrganizationFormReadOnly ? 'View' : 'Edit'} organization</PageHeading>
        ) : (
          <PageHeading>Create a new organization</PageHeading>
        )}
```

File: apps/web/frontend/components/organizationForm/organizationForm.tsx [hunk 14]
```diff
@@ -0,0 +1,92 @@
...
+        </button>
+        {organization?.canModify && (
+          <ArchiveOrUnarchiveOrganizationButton organization={organization} disabled={isSubmitting} />
+        )}
+        {!isOrganizationFormReadOnly && (
+          <button
+            className="btn btn-primary btn-sm"
+            type="submit"
+            disabled={isSubmitting}
+            title={isNewOrganization ? 'Create' : 'Save'}
```
Comment (PascalHelbig) [inline apps/web/frontend/components/organizationForm/organizationForm.tsx:82]:
```suggestion
```

File: apps/web/frontend/components/organizationForm/organizationForm.tsx [hunk 15]
```diff
@@ -0,0 +1,92 @@
...
+            form="organization-form"
+          >
+            {isNewOrganization ? 'Create' : 'Save'}
```
Comment (PascalHelbig) [inline apps/web/frontend/components/organizationForm/organizationForm.tsx:85]:
```suggestion
            {organization ? 'Save' : 'Create'}
```

File: apps/web/frontend/components/organizationTable.tsx [hunk 16]
```diff
@@ -0,0 +1,92 @@
...
+    #   name
+    # }
+  }
+`)
+
+interface OrganizationTableProps {
+  organizations: FragmentType<typeof OrganizationTableItemFragment>[]
+}
+
+// const MAX_NUMBER_OF_AVATARS = 5
```
Comment (PascalHelbig) [inline apps/web/frontend/components/organizationTable.tsx:21]:
```suggestion
  }
`)

interface OrganizationTableProps {
  organizations: FragmentType<typeof OrganizationTableItemFragment>[]
}
```

File: apps/web/frontend/components/organizationTable.tsx [hunk 17]
```diff
@@ -0,0 +1,92 @@
...
+      <table className="table">
+        <thead className="text-lg text-base-content">
+          <tr>
+            <th>Name</th>
+            {/* <th>Members</th> */}
+            {/* <th>Duration</th> */}
+          </tr>
+        </thead>
+        <tbody className="text-base">
+          {organizations.map((organization) => {
```
Comment (PascalHelbig) [inline apps/web/frontend/components/organizationTable.tsx:37]: remove comments

File: apps/web/pages/organizations/[id].page.tsx [hunk 18]
```diff
@@ -0,0 +1,82 @@
...
+import { ProtectedPage } from '../../frontend/components/protectedPage'
+import { graphql } from '../../frontend/generated/gql'
+import { OrganizationInput } from '../../frontend/generated/gql/graphql'
+
+const OrganizationQueryDocument = graphql(`
+  query organization($organizationId: ID!) {
+    organization(organizationId: $organizationId) {
+      id
+      ...OrganizationForm
+      #   ...ProjectMemberListProject
```
Comment (PascalHelbig) [inline apps/web/pages/organizations/[id].page.tsx:15]:
```suggestion
```

File: apps/web/pages/organizations/[id].page.tsx [hunk 19]
```diff
@@ -0,0 +1,82 @@
...
+    <ProtectedPage>
+      <OrganizationForm
+        organization={selectedOrganization}
+        onCancel={handleCancel}
+        onSubmit={handleSubmit}
+        hasError={!!organizationUpdateResult.error}
+      />
+      {/* <div role="tabpanel" className="tab-content rounded-box border-base-300 bg-base-100 p-6">
+          <ProjectMemberList project={selectedOrganization} />
+        </div> */}
```
Comment (PascalHelbig) [inline apps/web/pages/organizations/[id].page.tsx:77]: ✂️

File: apps/web/pages/organizations/index.page.tsx [hunk 20]
```diff
@@ -0,0 +1,54 @@
...
+  }
+
+  return (
+    <ProtectedPage>
+      <article>
+        <PageHeading>Organizations</PageHeading>
+        <div className="flex items-center justify-between gap-4">
+          <button className="btn btn-primary btn-sm" onClick={handleAddOrganization}>
+            New organization
+          </button>
```
Comment (PascalHelbig) [inline apps/web/pages/organizations/index.page.tsx:38]: use Link for routing

**Comment by PascalHelbig** [comment]
We also need GraphQL Tests for archive/unarchive organzation mutations

File: apps/web/frontend/components/organizationForm/archiveOrUnarchiveOrganizationButton/archiveOrganizationButton.tsx [hunk 21]
```diff
@@ -22,6 +22,7 @@ const OrganizationArchiveMutationDocument = graphql(`
 
 interface ArchiveOrganizationButtonProps {
   organization: FragmentType<typeof ArchiveOrganizationButtonFragment>
+  disabled: boolean
```
Comment (PascalHelbig) [inline apps/web/frontend/components/organizationForm/archiveOrUnarchiveOrganizationButton/archiveOrganizationButton.tsx:25]: Where is this used?

File: apps/web/frontend/components/organizationForm/archiveOrUnarchiveOrganizationButton/unarchiveOrganizationButton.tsx [hunk 22]
```diff
@@ -22,6 +22,7 @@ const OrganizationUnarchiveMutationDocument = graphql(`
 
 interface UnarchiveOrganizationButtonProps {
   organization: FragmentType<typeof UnarchiveOrganizationButtonFragment>
+  disabled: boolean
```
Comment (PascalHelbig) [inline apps/web/frontend/components/organizationForm/archiveOrUnarchiveOrganizationButton/unarchiveOrganizationButton.tsx:25]: see above

File: packages/backend/src/graphql/organization/mutations/organizationUpdateMutationField.test.ts [hunk 23]
```diff
@@ -0,0 +1,92 @@
...
+import { GraphQLError } from 'graphql'
+
+import { PrismaClient } from '@progwise/timebook-prisma'
+
+import { getTestServer } from '../../../getTestServer'
+
+const prisma = new PrismaClient()
+
+const organizationUpdateMutation = gql`
+  mutation organizationUpdateMutation($id: ID!, $data: OrganizationInput!) {
```
Comment (PascalHelbig) [inline packages/backend/src/graphql/organization/mutations/organizationUpdateMutationField.test.ts:11]:
```suggestion
  mutation organizationUpdate($id: ID!, $data: OrganizationInput!) {
```

**Review by PascalHelbig — APPROVED** [review]

**Pull request merged**